### PR TITLE
LiteLLM Proxy Server

### PR DESCRIPTION
**Pull request by amiraliemami**
Reviewer: @sidravi1 
Estimate: 1h 30mins

---

## Ticket

Fixes: [443](https://idinsight.atlassian.net/browse/AAQ-443?atlOrigin=eyJpIjoiMzI2MTY4ZjUyMWU4NGUwMmJjNzJmYTgzMjYxNDYzMjMiLCJwIjoiaiJ9)

### Goal

- Add the LiteLLM Proxy server to our stack

### Changes

- Added LiteLLM proxy server docker to our stack and updated:
  - Dev process (added Makefile targets and updated docs)
  - Docker Compose script (local deployment)
  - Added default configs
  - Added docs to "Other Components" section 
- Edited backend code to send all LLM and embeddings calls to the proxy server

Other updates:
- Docs cleanups
- Architecture diagram update (new figma page [here](https://www.figma.com/file/gvlztqJlZcaZTWTTjOCYDi/AAQ-Assets?type=design&node-id=443%3A502&mode=design&t=CLQt0cl3X1lhJr99-1))
- Updated embedding model to text-embedding-3-small for better performance at 5x less cost (see [OpenAI article](https://openai.com/blog/new-embedding-models-and-api-updates)) (see `litellm_config.yaml` files)
- Set default Guardrails model to GPT-3.5 turbo (see `litellm_config.yaml` files)

### Futue Tasks
- Update validation
- Update AWS terraform and CI/CD scripts
- Put custom alignscore behind this proxy
- Add Latest Updates blog post

## How has this been tested?
- Follow the docs/develop/setup.md steps and test endpoints
- Follow the docs/deployment/quick-setup.md steps and test deployment
- Follow the docs/develop/testing.md and run tests (including optional rails tests).

## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests (if applicable)
- [x] I have updated the requirements (if applicable)
- [x] I have updated the README file (if applicable)
- [x] I have updated affected documentation (if applicable)

### CODE REVIEW

File: litellm_proxy/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# LiteLLM Proxy Docker
+
+All of our embeddings and LLM calls will now pass through this proxy. See [confluence page](https://idinsight.atlassian.net/wiki/spaces/PD/pages/2452488357/LiteLLM+Proxy+Server) for details.
```
Comment (suzinyou) [inline litellm_proxy/README.md:3]: This is a private document :o

File: litellm_proxy/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+## How to run
+
+1.  Pull LiteLLM proxy image
+
+        docker pull ghcr.io/berriai/litellm:main-v1.34.6
```
Comment (suzinyou) [inline litellm_proxy/README.md:9]: Will you add this as another service on docker-compose?

File: litellm_proxy/README.md (hunk 1 above)
Comment (amiraliemami) [reply litellm_proxy/README.md:3]: Yeah it's temporary for the draft PR - will remove.

File: litellm_proxy/README.md (hunk 2 above)
Comment (amiraliemami) [reply litellm_proxy/README.md:9]: yup

File: core_backend/app/llm_call/parse_input.py
Comment (suzinyou) [inline core_backend/app/llm_call/parse_input.py:1]: amazing

File: docs/deployment/architecture.md [hunk 3]
```diff
@@ -1,13 +1,7 @@
...
 
 <p align="center">
-  <img src="../../images/architecture-docker.png" alt="Architecture"/>
-</p>
-
-A reverse proxy manages all incoming traffic to the service. The database is only accessed by the core app.
-
-<p align="center">
-  <img src="../../images/architecture-traffic.png" alt="Flow"/>
+  <img src="../../images/architecture.png" alt="Flow"/>
```
Comment (sidravi1) [inline docs/deployment/architecture.md:6]: /nit Can you move the "LiteLLM Proxy" icon down to be in a similar location as Postgres?

File: docs/develop/setup.md [hunk 4]
```diff
@@ -1,121 +1,106 @@
 # Setting up your development environment
 
+!!! warning "You need to have installed [Docker](https://docs.docker.com/get-docker/)"
+
 ## Option 1 - Using Docker Compose Watch
 
-This option uses the same Docker Compose script as deployment and so is *environment-agnostic*. It's good for
-quick startup and end-to-end testing but the downside is that changes take 5-10s to be reflected in the app (1).
-{ .annotate }
+Recommended.
```
Comment (sidravi1) [inline docs/develop/setup.md:7]: Not sure we need this.

File: docs/develop/setup.md [hunk 5]
```diff
@@ -1,121 +1,106 @@
...
 
-1. This is because the backend and admin app images are rebuilt every time there is a change to their respective code folders.
+| Pros | Cons |
+| --- | --- |
+| Good for end-to-end testing | Changes take 5-10s to be reflected in the app |
```
Comment (sidravi1) [inline docs/develop/setup.md:11]: Local env identical to production deployment

File: docs/develop/setup.md [hunk 6]
```diff
@@ -1,121 +1,106 @@
...
+| Environment-agnostic | |
```
Comment (sidravi1) [inline docs/develop/setup.md:12]: What does this mean?

File: docs/develop/setup.md [hunk 7]
```diff
@@ -1,121 +1,106 @@
...
+| Set environemnt variables and configs once | |
```
Comment (sidravi1) [inline docs/develop/setup.md:13]:
```suggestion
| Set environment variables and configs once | |
```

File: docs/develop/setup.md [hunk 8]
```diff
@@ -1,121 +1,106 @@
...
-3. run
 
-        docker compose -f docker-compose.yml -f docker-compose.dev.yml -p aaq-stack watch
-
-The app will now run and update with any changes made to the `core_backend` or `admin_app` folders.
+2. copy `template.env` to a new file `.env` and set the necessary variables (for local deployment, you just
+need to set the `OPENAI_API_KEY` and can leave everything else as default)
 
-The admin app will be available on `http://localhost` and the backend API testing UI on `http://localhost/api/docs`.
+3. (optional) Edit which LLMscon are used in the `litellm_config.yaml`
```
Comment (sidravi1) [inline docs/develop/setup.md:22]:
```suggestion
3. (optional) Edit which LLMs are used in the `litellm_config.yaml`
```

Comment (sidravi1) [reply docs/develop/setup.md:22]: What if we removed this line?

File: docs/develop/setup.md [hunk 9]
```diff
@@ -1,121 +1,106 @@
...
-Takes a little more configuration before each run.
+        docker compose -f docker-compose.yml -f docker-compose.dev.yml -p aaq-stack watch
 
-Once you have your local environment setup (1), you can setup the components needed to
-develop your new feature.
-{ .annotate }
+The app will now run and update with any changes made to the `core_backend` or `admin_app` folders.
 
-1. if you haven't see [Contributing to AAQ]("./contributing.md")
+The admin app will be available on [http://localhost](http://localhost) and the backend API testing UI on [http://localhost/api/docs](http://localhost/api/docs).
```
Comment (sidravi1) [inline docs/develop/setup.md:30]: not https?

File: docs/develop/setup.md [hunk 10]
```diff
@@ -1,121 +1,106 @@
...
 
-You can stop and remove the PostgreSQL container using:
+2. Set required environment variables in your terminal using
 
-    make teardown-db
+        export PROMETHEUS_MULTIPROC_DIR=/tmp
+        export OPENAI_API_KEY=sk...
 
-Otherwise, you can run them manually as below.
+3. In the root of the repository, edit the `litellm_config.yaml` as required.
```
Comment (sidravi1) [inline docs/develop/setup.md:52]: Can you make this the same language as 3 in the previous one?

File: docs/develop/setup.md [hunk 11]
```diff
@@ -1,121 +1,106 @@
...
-    preferable to create your database from scratch each time. But if you wish to persist data
-    use a volume as below:
+        python core_backend/main.py
 
-        docker run --name postgres-local \
-        -e POSTGRES_PASSWORD=<your password> \
-        -p 5432:5432 \
-        -v postgres_local_vol:/var/lib/postrges/data \
-        -d pgvector/pgvector:pg16
+    This will launch the application in "reload" mode i.e. the app with automatically
```
Comment (sidravi1) [inline docs/develop/setup.md:62]:
```suggestion
    This will launch the application in "reload" mode i.e. the app will automatically
```

**Review by sidravi1 — APPROVED** [review]
Thanks @amiraliemami - Looks good! Some minor changes.
Also, weird that CI/CD tests didn't run for this. Can you look into it?

**Review by sidravi1 — APPROVED** [review]
Ok fixed the image issue. This should be good to go!

File: docs/develop/setup.md (hunk 4 above)
Comment (amiraliemami) [reply docs/develop/setup.md:7]: deleted

File: docs/develop/setup.md (hunk 9 above)
Comment (amiraliemami) [reply docs/develop/setup.md:30]: fixed (both work due to redirection)

**Pull request merged**